kimchi/src/folding/mod.rs

```diff
@@ -0,0 +1,319 @@
+use self::error::compute_error;
+use crate::circuits::{
+    expr::{CacheId, Expr, FeatureFlag},
+    gate::GateType,
+};
+use ark_ff::{Field, One, Zero};
+use ark_poly::{EvaluationDomain, Evaluations, Radix2EvaluationDomain};
+use expressions::{folding_expression, IntegratedFoldingExpr};
+pub use instance_witness::{InstanceTrait, RelaxedInstance, RelaxedWitness, WitnessTrait};
+use instance_witness::{RelaxableInstance, RelaxablePair};
+use poly_commitment::PolyComm;
+use poly_commitment::{commitment::CommitmentCurve, SRS};
+use std::{collections::HashMap, marker::PhantomData, sync::Arc};
+
+mod error;
+mod expressions;
+mod instance_witness;
+#[cfg(test)]
+mod test;
+
+pub(crate) trait FoldingEnv<F> {
+    fn witness(&self, i: usize, side: bool) -> &Vec<F>;
+    fn witness_ext(&self, i: usize, side: bool) -> &Vec<F>;
+    fn index(&self, i: &GateType) -> &Vec<F>;
+    fn coefficient(&self, i: usize) -> &Vec<F>;
+    fn shift(&self) -> &Vec<F>;
+    fn lagrange_basis(&self, i: &i32) -> Vec<F>;
+    ///a vec of just zeros of the same length as other columns
+    fn zero_vec(&self) -> Vec<F>;
+}
+
+pub(crate) enum EvalLeaf<'a, F: Field> {
+    Const(F),
+    Col(&'a Vec<F>),
+    Result(Vec<F>),
+}
+
+impl<'a, F: Field> EvalLeaf<'a, F> {
+    fn map<M: Fn(&F) -> F, I: Fn(&mut F)>(self, map: M, in_place: I) -> Self {
+        use EvalLeaf::*;
+        match self {
+            Const(c) => Const(map(&c)),
+            Col(col) => {
+                let res = col.iter().map(map).collect();
+                Result(res)
+            }
+            Result(mut col) => {
+                for cell in col.iter_mut() {
```

Comment:
If you want to apply functional programming, why not using `for_each` instead of the evil `for`? :)

Reply:
I didn't want to apply anything in particular, generally I prefer `for in` for doing side effects or mutations with each element, and I find it clearer than `for_each` unless I already have a closure available.